x/stakeibc/keeper/unbonding_records.go

```diff
@@ -43,171 +69,261 @@ func (k Keeper) CreateEpochUnbondingRecord(ctx sdk.Context, epochNumber uint64)
-	if len(msgs) == 0 {
-		return nil, sdkmath.ZeroInt(), nil, nil, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "Target unbonded amount was 0 for each validator")
+	// Sanity check that we had enough capacity to unbond
+	if !remainingUnbondAmount.IsZero() {
+		return msgs, unbondings,
+			fmt.Errorf("unable to unbond full amount (%v) from %v", totalUnbondAmount, hostZone.ChainId)
 	}
 
-	// Store the callback data
-	undelegateCallback := types.UndelegateCallback{
-		HostZoneId:              hostZone.ChainId,
-		SplitDelegations:        splitDelegations,
-		EpochUnbondingRecordIds: epochUnbondingRecordIds,
+	return msgs, unbondings, nil
+}
+
+// Submits undelegation ICA messages for a given host zone
+//
+// The total unbond amount is determined from the epoch unbonding records
+// The unbond amount is then distributed amongst validators proportional to how unbalanced they are.
+// As a result, unbondings lead to a more balanced distribution of stake across validators
+//
+// Context:
+//   Over time, as LSM Liquid stakes are accepted, the total stake managed by the protocol becomes unbalanced
+//   as liquid stakes are not aligned with the validator weights. This is only rebalanced once per unbonding period
+func (k Keeper) UnbondFromHostZone(ctx sdk.Context, hostZone types.HostZone) error {
+	k.Logger(ctx).Info(utils.LogWithHostZone(hostZone.ChainId,
+		"Preparing MsgUndelegates from the delegation account to each validator"))
+
+	// Confirm the delegation account was registered
+	if hostZone.DelegationIcaAddress == "" {
+		return errorsmod.Wrapf(types.ErrICAAccountNotFound, "no delegation account found for %s", hostZone.ChainId)
 	}
-	k.Logger(ctx).Info(utils.LogWithHostZone(hostZone.ChainId, "Marshalling UndelegateCallback args: %+v", undelegateCallback))
-	marshalledCallbackArgs, err = k.MarshalUndelegateCallbackArgs(ctx, undelegateCallback)
-	if err != nil {
-		k.Logger(ctx).Error(err.Error())
-		return nil, sdkmath.ZeroInt(), nil, nil, errorsmod.Wrap(sdkerrors.ErrNotFound, err.Error())
+
+	// Iterate through every unbonding record and sum the total amount to unbond for the given host zone
+	totalAmountToUnbond, epochUnbondingRecordIds := k.GetTotalUnbondAmountAndRecordsIds(ctx, hostZone.ChainId)
+	k.Logger(ctx).Info(utils.LogWithHostZone(hostZone.ChainId,
+		"Total unbonded amount: %v%s", totalAmountToUnbond, hostZone.HostDenom))
+
+	// If there's nothing to unbond, return and move on to the next host zone
+	if totalAmountToUnbond.IsZero() {
+		return nil
 	}
 
-	return msgs, totalAmountToUnbond, marshalledCallbackArgs, epochUnbondingRecordIds, nil
-}
+	// Determine the ideal balanced delegation for each validator after the unbonding
+	//   (as if we were to unbond and then rebalance)
+	// This will serve as the starting point for determining how much to unbond each validator
+	currentTotalDelegation := k.GetTotalValidatorDelegations(hostZone)
+	delegationAfterUnbonding := currentTotalDelegation.Sub(totalAmountToUnbond)
```

Comment:
It is so simple now and makes good reuse of functions originally defined for the rebalance, this might be some of my favorite code in the codebase.  If we make the liquidstake delegate code also dynamically move things towards the more balanced state (maybe doing this same exact algorithm in reverse) instead of just blindly splitting according to the weights I think we actually improved the aethetics and the average performance of the protocol at the same time!

Reply:
I agree! Really great work coming up with this algorithm!